src/probe/probe.c

```diff
@@ -15,42 +22,529 @@
+
+#define PROBE_BUFFER_LOCAL_SIZE		8192
+
+struct probe_dma_buf {
+	uintptr_t w_ptr;
+	uintptr_t r_ptr;
+	uintptr_t addr;
+	uintptr_t end_addr;
+	uint32_t size;
+	uint32_t avail;
+};
+
+struct probe_dma_ext {
+	uint32_t stream_tag;
+	uint32_t dma_buffer_size;
+	struct dma_sg_config config;
+	struct probe_dma_buf dmapb;
+	struct dma_copy dc;
+};
+
+struct probe_pdata {
+	struct probe_dma_ext ext_dma;
+	struct probe_dma_ext inject_dma[CONFIG_PROBE_DMA_MAX];
+	struct probe_point probe_points[CONFIG_PROBE_POINTS_MAX];
+};
+
+static struct probe_pdata *_probe;
+
+static int probe_dma_buffer_init(struct probe_dma_buf *buffer, uint32_t size)
+{
+	/* allocate new buffer */
+	buffer->addr = (uintptr_t)rballoc_align(SOF_MEM_ZONE_BUFFER,
+						SOF_MEM_CAPS_DMA, size, 128);
+
+	if (!buffer->addr) {
+		trace_probe_error("probe_dma_buffer_init() error: alloc failed");
+		return -ENOMEM;
+	}
+
+	bzero((void *)buffer->addr, size);
+	dcache_writeback_region((void *)buffer->addr, size);
+
+	/* initialise the DMA buffer */
+	buffer->size = size;
+	buffer->w_ptr = buffer->addr;
+	buffer->r_ptr = buffer->addr;
+	buffer->end_addr = buffer->addr + buffer->size;
+	buffer->avail = 0;
+
+	return 0;
+}
+
+static int probe_dma_init(struct probe_dma_ext *dma)
+{
+	struct dma_sg_config config;
+	uint32_t elem_size, elem_addr, elem_num;
+	int err = 0;
+
+	/* initialize dma buffer */
+	err = probe_dma_buffer_init(&dma->dmapb, PROBE_BUFFER_LOCAL_SIZE);
+	if (err < 0)
+		return err;
+
+	/* request HDA DMA in the dir LMEM->HMEM with shared access */
+	dma->dc.dmac = dma_get(DMA_DIR_LMEM_TO_HMEM, 0, DMA_DEV_HOST,
+			       DMA_ACCESS_SHARED);
+	if (!dma->dc.dmac) {
+		trace_probe_error("probe_dma_init() error: dma->dc.dmac = NULL");
+		return -ENODEV;
+	}
+
+	err = dma_copy_set_stream_tag(&dma->dc, dma->stream_tag);
+	if (err < 0)
+		return err;
+
+	elem_size = sizeof(uint64_t) * 32;
+	elem_addr = (uint32_t)dma->dmapb.addr;
+	elem_num = PROBE_BUFFER_LOCAL_SIZE / elem_size;
+
+	config.direction = DMA_DIR_LMEM_TO_HMEM;
+	config.src_width = sizeof(uint32_t);
+	config.dest_width = sizeof(uint32_t);
+	config.cyclic = 0;
+
+	err = dma_sg_alloc(&config.elem_array, SOF_MEM_ZONE_RUNTIME,
+			   config.direction, elem_num, elem_size, elem_addr, 0);
+	if (err < 0)
+		return err;
+
+	err = dma_set_config(dma->dc.chan, &config);
+	if (err < 0)
+		return err;
+
+	dma_sg_free(&config.elem_array);
+
+	err = dma_start(dma->dc.chan);
+	if (err < 0)
+		return err;
+
+	return 0;
+}
+
+static int probe_dma_deinit(struct probe_dma_ext *dma)
+{
+	int err = 0;
+
+	err = dma_stop(dma->dc.chan);
+	if (err < 0) {
+		trace_probe_error("probe_dma_deinit() error: dma_stop() failed");
+		return err;
+	}
+
+	dma_channel_put(dma->dc.chan);
+
+	rfree((void *)dma->dmapb.addr);
+	dma->dmapb.addr = 0;
+
+	dma->stream_tag = PROBE_DMA_INVALID;
+
+	return 0;
+}
+
 int probe_init(struct probe_dma *probe_dma)
 {
+	uint32_t i;
+	int err;
+
+	tracev_probe("probe_init()");
+
+	if (_probe) {
```

Comment:
This should be in teh main sof context, @tlauda has been working on this lately please align.

Reply:
done